pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+        this_data = analysis[val]
+        trace = go.Scatter(
+            x = x_points,
+            y = this_data,
+            name = val,
+            line = dict(
+                color = ('rgb(%d, %d, %d)' % (r(), r(), r())),
+                width = 4)
+        )
+        data.append(trace)
+        
+        # Find the polynomial runtime
+        runtime[val] = get_runtime_approximate(x_points, this_data)
+        #print str(val) + ": " + str(runtime[val][0]) + "n^" + str(runtime[val][1])
+        
+    if bounds is not None:
+        for name in bounds.keys():
+            # Create the graph
+            function = bounds[name]
+            this_data = [function(i) for i in x_points]
+            trace = go.Scatter(
+                x = x_points,
+                y = this_data,
+                name = name,
+                line = dict(
+                    color = ('rgb(%d, %d, %d)' % (r(), r(), r())),
+                    width = 4,
+                    dash = 'dash')
+            )
+            data.append(trace)
+        
+
+    # Edit the layout
+    layout = dict(title = title,
+                  xaxis = dict(title = x_axis),
+                  yaxis = dict(title = y_axis),
+                  )
+
+    fig = dict(data=data, layout=layout)
+    plotly.offline.plot(fig)
+
+def aos_to_soa(aos):
+        '''
+        Converts an Array of Structures into a Structure of Arrays. Not robust yet, as
+        the keys to be used in the structure are determined from the first structure in
+        the input Array of Structure
+        :param aos: An Array of Structure to convert into a Structure of Arrays
+        '''
+        
+        result = {}
+        for key in aos[0]:
+            result[key] = []
+            
+        for test in aos:
+            for item in test.keys():
+                val = test[item]
+                result[item].append(val)
+                
+        return result
+
+def get_runtime_approximate(x, y):
+    '''
+    Given x and y points, determines an approximate running
+    time for the given points using fitting
+    '''
+    
+    # Check if this is polynomial
+    current_max = 0
+    poly_result = np.polyfit(x, y, 10)
+    degree = 10
+    polynomial_degree = (None, None)
+    while degree > 0:
+        degree -= 1
+        coeff = poly_result[degree]
+        if coeff > current_max:
+            current_max = coeff
+            polynomial_degree = (coeff, 10 - degree)
+            
+    return polynomial_degree
+        
+    # Check if this is exponential
+    
+    # Check if this is logarithmic
+    
+    
+def get_all_gate_types():
+    '''
+    Simply returns a set of all possible Quil gate types
+    '''
+    return gates.STANDARD_GATES.keys() + \
+            ['RESET', 'NOP', 'WAIT', 'HALT', 'TRUE', \
+             'FALSE', 'NOT', 'AND', 'OR', 'WAIT', 'MOVE', \
+             'EXCHANGE', 'JUMP-WHEN', 'JUMP-UNLESS', 'DEFGATE', 'MEASURE']
+    
+    
+if __name__ == "__main__":
```

Comment:
it's great that you have tests here that you are using while developing. You may as well put them into an `test_analysis.py` file and get a head start on having tested code.